worker/echo/main.go

```diff
@@ -15,19 +16,40 @@ import (
 )
 
 var sleepTime time.Duration
+var cancelChan map[string]chan struct{}
+var chanMutex = &sync.RWMutex{}
 
 // echo opens a new dbus connection and calls the
 // com.redhat.Yggdrasil1.Dispatcher1.Transmit method, returning the metadata and
-// data it received.
+// data it received. Cancelled messages will only be handled if the worker is running
+// in the slow mode.
 func echo(w *worker.Worker, addr string, id string, responseTo string, metadata map[string]string, data []byte) error {
 	if err := w.EmitEvent(ipc.WorkerEventNameWorking, fmt.Sprintf("echoing %v", data)); err != nil {
 		return fmt.Errorf("cannot call EmitEvent: %w", err)
 	}
 
 	// Sleep time between receiving the message and sending it
 	if sleepTime > 0 {
-		log.Infof("sleeping: %v", sleepTime)
+		// Setting the channel to handle cancelation
+		chanMutex.Lock()
+		cancelChan[id] = make(chan struct{})
+		chanMutex.Unlock()
+
+		log.Tracef("sleeping: %v", sleepTime)
 		time.Sleep(sleepTime)
+
+		// Cancel message if it has been sent a cancel message
+		// during sleep time
+		chanMutex.Lock()
+		select {
+		case <-cancelChan[id]:
+			log.Tracef("canceled echo message id: %v", id)
+			delete(cancelChan, id)
+			return nil
+		default:
+			delete(cancelChan, id)
+		}
+		chanMutex.Unlock()
```

Comment:
Again this looks like as something that will be duplicate in all workers. Shouldn't we create something that will help all developers?

Reply:
As previously mentioned, channels are specific to Golang, and this implementation can be different depending on each worker. How to handle the cancellation should be up to each worker.